### ISSUE TITLE
[Feature Request] `field` keyword

### ISSUE THREAD

**Issue by avivbeeri** [wren-lang/wren#849]
I imagine this will get rejected, but I need to get it out of my head.

When I'm working on projects in Wren, I find myself making a lot of fields for classes with symmetric getters and setters, in this sort of style:

```
class Entity {
  construct (pos) {
    _pos = pos
  }

 pos { _pos }
 pos=(v) { pos = v }
}
```

It would be nice if there was a keyword sugar to generate this more conveniently.

For example:
```
class Entity {
  construct (pos) {
    _pos = pos
  }

  field pos
}
```

I think you could also do this for static fields: `static field pos`

Obviously there are downsides to adding an extra keyword to the language, i'm not sure if there's a way to fix that in a backwards compatible way.

**Comment by ChayimFriedman2** [comment]
>  i'm not sure if there's a way to fix that in a backwards compatible way.
There is, using contextual keywords, but this will complicate both the lexer and the parser a lot. Not worth it IMHO.

**Comment by ChayimFriedman2** [comment]
I think this will be better with the following syntax:
```wren
class Entity {
  construct (pos) {
    _pos = pos
  }

  pos = _pos
}
```
It will also be backwards compatible.

**Comment by mhermier** [comment]
The complexity is not very high, there are a few places where we create asm
stubs by hand and it's not really complicated.

The syntax should be investigated a little bit more though.

It is true that you use a more data oriented approach you need a lot of
trivial getters/setters.

**Comment by avivbeeri** [comment]
I believe ruby has an implementation of a feature like this in the works already, with a less problematic syntax

**Comment by billyquith** [comment]
Hi @avivbeer. What is the benefit to using that style, as you say? I.e. hiding your member variables, but then exposing them as properties? That might be something you do idiomatically in C++ (but you can set access). What is the reason for doing it in Wren? This is fine:

	class Entity {
	  construct new (pos) {
	   pos = pos
	  }
	}

Then, later, if you need to hide `pos` you can rename the hidden version. That way, you only do the work for the variables that have a property interface.

	class Entity {
	  construct new (pos) {
	   _pos = pos
	  }

	 pos { _pos }
	 pos=(v) { _pos = v }
	}

Is there a reason not to change that style? I'm pretty new to Wren so I'm still learning it myself.

**Comment by avivbeeri** [comment]
Your first example there doesn't actually achieve anything: The pos value will just be lost since you assign it to itself.
Assuming you meant to write: 
```
class Entity {
  construct new (pos) {
    _pos = pos // this creates a private _pos field
  }
}
```

Then the `_pos` is a private value. If it's your only value, manually creating getters and setters is fine.

But for certain applications, you end up having lots of variables, all with public getters and setters, and it gets _very_ repetitive.
 
For example, in DOME: https://github.com/domeengine/dome/blob/fa067c51cbff3ae4087b8acb681c9507aca58fd4/src/modules/audio.wren#L133
```
  volume { _volume }
  volume=(volume) { _volume = volume }
  loop { _loop }
  loop=(loop) { _loop = loop }
  pan { _pan }
  pan=(pan) { _pan = pan }
```

It would be nice if I didn't have to declare each getter and setter manually, when I'm creating identical methods for each one.

In some game projects, you could potentially end up with 6-7 of these, which is both tedious, and error prone

**Comment by billyquith** [comment]
Ah, ok, `_` is mandatory. 🍏 Thanks. I thought it was stylistic like Python. In that case I think it might be best to add this without adding a keyword. Wren seems to try to be very concise. Maybe:

	class Entity {
	  construct new (pos) {
	   _pos = pos
	  }

	  pos = _pos
	}

**Comment by avivbeeri** [comment]
Indeed. As mentioned above, a potential way of doing this in the works, which doesn't add new keywords, so we can look forward to that.

**Comment by mhermier** [comment]
I found `foo = _foo` noisy for something that could be automated, unless it
is a rename

**Comment by PureFox48** [comment]
A problem with having a combined getter/setter syntax in a dynamically typed language such as Wren is that you can't then type-check the incoming setter value.

For example, I have a lot of code which follows this sort of pattern:

```javascript
foo { _foo }

foo=(f) {
    if (!(f is Foo)) Fiber.abort("Argument must be a Foo.")
    _foo = f
}
```
Sometimes rather than aborting I'll just change the type of the argument to something more suitable, stringifying a Num say if I'm expecting a String.

It's rare for me not to care about the type of the incoming value so, even if we had a syntax for this, I'm not sure how much use it would be when writing robust production code.

Incidentally, one thing I would find useful is an  ```isnot``` operator to cut down on the parentheses in the above code.

**Comment by mhermier** [comment]
This is not a replacement for the existing syntax. And this can be used
until the urge for specific types are required. Thought I would not
encourage type checking at all, unless really required.

In your example, I would change the flow of `if` to positive, so there is
only a final default handling and possibly a single `Fiber.abort()`. Not
that it should drastically change performance, but I find it to be nice
flow with only one error handling at the end.

**Comment by PureFox48** [comment]
I realize that this wouldn't be a replacement for what we do already, just another option. As such I don't really have a problem with it even though I'd rarely use it myself.

I guess it's a matter of taste how you arrange the code and sometimes I find it easier to do it the way you suggested.  But the Wren code in ```wren_core.wren``` does at least seem to do it the way I usually do and get the type checking out of the way first.

**Comment by clsource** [comment]
This reminds me of `@synthesize` in Objective-C
https://programmer.help/blogs/objective-c-property-synthesize-and-dynamic.html

> Objc recommends that we access the properties of an object through the set/get method. Obviously, manually adding the declaration and implementation of set/get methods for each attribute is a very low cost-effective duplication of effort. Therefore, objc provides some keywords to help us simplify this process. That's exactly what it is.

`@synthesize` specifies the implementation of the set/get method. By default, `@synthesize` generates a member variable with the same name as the target of the set/get.

I don't know if this could be added in as a method?

```js
class Entity {
  pos {_post}

  synthesize() {
    // Will generate set/get methods for these properties
    return [_post]
  }
  construct (pos) {
    _pos = pos
    this.synthesize()
  }
}
```

**Comment by mhermier** [comment]
It is not possible for multiple reasons for now. And in addition I find it
not safe that it can be changed dynamicaly on each invocation. For safety
it should be a grammar modification.

**Comment by ChayimFriedman2** [comment]
I would like to add that changing classes dynamically go against the philosophy of Wren:

 > Wren is strictly class-based. Every object is an instance of a class. Classes in turn have a well-defined declarative syntax, and cannot be imperatively modified. In addition, fields in Wren are private to the class—they can only be accessed from methods defined directly on that class.

**Comment by clsource** [comment]
In that case then using `synthesize` as a keyword (similar to static) maybe?

```js
class Entity {
  synthesize pos
  static synthesize message
  construct (pos) {
    this.pos = pos
  }
}
```

**Comment by ChayimFriedman2** [comment]
I like the idea. It like the proposed `field` keyword but less boilerplate-y.

But maybe we want an other keyword.

**Comment by mhermier** [comment]
`var` would be a good candidate, thought I would reserve it for future 
to allow non '_' prefixed member names.

**Comment by ChayimFriedman2** [comment]
I don't think we want non `_` prefixed field names. That's redundant IMO. I would vote for `property`.

BTW, you don't even have a name for the constructor and you don't even need `this.` if you rename the parameter:
```wren
class Entity {
  property pos
  construct bew(position) {
    pos = position
  }
}
```

**Comment by mhermier** [comment]
`this` is not required as long as you declare your properties/field/... 
before hand. Otherwise I think name resolution kick outside of the class 
scope.

**Comment by ChayimFriedman2** [comment]
I'm not sure I understand you correctly, but if you mean that the compiler can't find methods that declared after reference, you're wrong.

**Comment by mhermier** [comment]
It does work unless you have an external variable of the same name. The
upper scope resolution kicks in. So you have either to move declaration
first or use this.

**Comment by clsource** [comment]
Other ideas would be 

```js
class Entity {
  // Only synthezise getter
  $pos
  // Synthezise getter and setter
  static $message$
  construct (pos) {
    pos = pos
  }
}
```

**Comment by ChayimFriedman2** [comment]
> It does work unless you have an external variable of the same name. The upper scope resolution kicks in. So you have either to move declaration first or use this.

Still I'm not sure I understand you but if you mean for example:
```wren
var a = "Outer"
class C {
  f() {
    System.print(a)
  }
  a { "Method" }
}
```
Then that depends on the captialness of the name.

@clsource I think this is slightly out of scope. Getter+setter combination is common, only getter/only setter are not common nor boilerplatey.

**Comment by clsource** [comment]
How about this syntax?
Just using the available `is` and `in` keywords

```js
class Entity {
  pos is in
  construct (pos) {
    pos = pos
  }
}
```

**Comment by ChayimFriedman2** [comment]
First, in the constructor, `this.pos = pos`.

Second, redundant. A keyword `property` is the best IMO.

**Comment by clsource** [comment]
Ok here is more elaborate.

```js
class Entity {
  message {in}
  construct new(message) {
    this.message = message
  }
}


var hello = Entity.new("world")
System.print(hello.message)
```

**Comment by ChayimFriedman2** [comment]
Why not just
```wren
class Entity {
  property message
  construct new(msg) {
    message = msg
  }
}

var hello = Entity.new("world")
System.print(hello.message)
```
Clearer, doesn't it?

**Comment by clsource** [comment]
But it would require a new keyword to be created.
My guess is that this feature have more possibilities to be included if does not requires a new keyword and uses the already available keywords.

I think that the keywords: `is`,  `as`, `this` and `in` can be used to form a phrase that denotes an understandable meaning.

```js
class Entity {
  message in this
  construct new(message) {
    this.message = message
  }
}


var hello = Entity.new("world")
System.print(hello.message)
```

**Comment by ChayimFriedman2** [comment]
Let's ask @ruby0x1 about that.

**Comment by mhermier** [comment]
I was thinking that `var` could be the correct word, but then I thought
about private member aliasing. And since the natural way to declare it,
would be to use `=` I think it could be confusing.

We need a keyword of some sort, so the grammar stays LL(1). I'm not a big
fan of post keywords since it will make aliasing quite complicated.

**Comment by ChayimFriedman2** [comment]
I prefer `property` but I understand the wish to not add a keyword.
